src/blockchain_utilities/blockchain_blackball.cpp

```diff
@@ -0,0 +1,403 @@
+  tools::ringdb ringdb(output_file_path.string());
+
+  for (size_t n = 0; n < inputs.size(); ++n)
+  {
+    LOG_PRINT_L0("Reading blockchain from " << inputs[n]);
+    for_all_transactions(inputs[n], [&](const cryptonote::transaction_prefix &tx)->bool
+    {
+      for (const auto &in: tx.vin)
+      {
+        if (in.type() != typeid(txin_to_key))
+          continue;
+        const auto &txin = boost::get<txin_to_key>(in);
+        if (opt_rct_only && txin.amount != 0)
+          continue;
+
+        const std::vector<uint64_t> absolute = cryptonote::relative_output_offsets_to_absolute(txin.key_offsets);
+        if (n == 0)
+          for (uint64_t out: absolute)
+            outputs[output_data(txin.amount, out)].insert(txin.k_image);
+
+        const uint32_t ring_size = txin.key_offsets.size();
+        if (ring_size == 1)
+        {
+          const crypto::public_key pkey = core_storage[n]->get_output_key(txin.amount, txin.key_offsets[0]);
+          MINFO("Blackballing output " << pkey << ", due to being used in a 1-ring");
+          ringdb.blackball(pkey);
+          newly_spent.insert(output_data(txin.amount, txin.key_offsets[0]));
+          spent.insert(output_data(txin.amount, txin.key_offsets[0]));
+        }
+        else if (relative_rings.find(txin.k_image) != relative_rings.end())
+        {
+          MINFO("Key image " << txin.k_image << " already seen: rings " <<
+              boost::join(relative_rings[txin.k_image] | boost::adaptors::transformed([](uint64_t out){return std::to_string(out);}), " ") <<
+              ", " << boost::join(txin.key_offsets | boost::adaptors::transformed([](uint64_t out){return std::to_string(out);}), " "));
+          if (relative_rings[txin.k_image] != txin.key_offsets)
+          {
+            MINFO("Rings are different");
+            const std::vector<uint64_t> r0 = cryptonote::relative_output_offsets_to_absolute(relative_rings[txin.k_image]);
+            const std::vector<uint64_t> r1 = cryptonote::relative_output_offsets_to_absolute(txin.key_offsets);
+            std::vector<uint64_t> common;
+            for (uint64_t out: r0)
+            {
+              if (std::find(r1.begin(), r1.end(), out) != r1.end())
+                common.push_back(out);
+            }
+            if (common.empty())
+            {
+              MERROR("Rings for the same key image are disjoint");
+            }
+            else if (common.size() == 1)
+            {
+              const crypto::public_key pkey = core_storage[n]->get_output_key(txin.amount, common[0]);
+              MINFO("Blackballing output " << pkey << ", due to being used in rings with a single common element");
+              ringdb.blackball(pkey);
+              newly_spent.insert(output_data(txin.amount, common[0]));
+              spent.insert(output_data(txin.amount, common[0]));
+            }
+            else
+            {
+              MINFO("The intersection has more than one element, it's still ok");
```

Comment:
If `r0` and `r1` are of different lengths, ring members that aren't common in both get identified as fakes. To make use of this information, `relative_rings[txin.k_image]` should be set to the smallest ring among all that share the same key image.

Reply:
Very good observation, thanks.

Reply:
> should be set to the smallest ring among all that share the same key image.

Actually, not just the smallest ring among all, but the set of common members which may be smaller than any of the relevant rings.

Reply:
That's what I've done, yes.